### PR TITLE
Use full duplex in tpuart_data_link_layer.

### PR DESCRIPTION
**Pull request by mptei**
tl;dr: PR implements full duplex in tpuart layer. Stay in loop in time critical handling.

The tpuart-layer is implemented in half duplex with a common state for sending and receiving. This works only with very low bus load and leads to many transmission errors. The tpuart communication is full duplex. While uploading a layer-2 packet to the TPUART it may receive packets from the bus. Also it has to acknowledge the received packet as soon as possible while uploading.

Commit 'TPUART full duplex handling' changes the tpuart-layer to be full duplex. For that it uses two state machines. One for sending and one for receiving. It also recognizes end-of-packet (EOP) gaps. The TPUART signals the end of a packet with such a gap. This must be used to stop processing a layer-2 packet receiption and can be used to resynchronize to the stream.

I have (stress) tested this PR with success. As long as the loop is called fast enough the communication is very stable. There are no lost packets.

If the load increases and the loop is called not fast enough it came to many receiption errors. The problem here is that the receiption of a layer-2 packet is very time critical. The EOP is only 2-2.5ms wide. To deal with this, the commit 'Stay in loop when RX_L_DATA.' stays in loop as long as a layer-2 receiption is ongoing. This makes receiption stable on higher load.

I have tested this PR against a TPUART2 and an NCN5120. 

Errors as seen here: https://github.com/thelsing/knx/issues/89#issuecomment-697998765 or https://github.com/thelsing/knx/issues/83#issue-686347861 are solved with this PR.

The pattern:
```
expected L_DATA_CON not received
got U_CONFIGURE_IND: 0x11
got UNEXPECTED: 0x2
got L_ACKN_IND
got L_ACKN_IND
got UNEXPECTED: 0xE1
got U_CONFIGURE_IND: 0x1
got L_ACKN_IND
got UNEXPECTED: 0xDC
got U_STATE_IND: 0x17
got unexpected L_DATA_CON
```
means nothing else, that when the state machine is waiting for L_DATA_CON another layer-2 packet rushes in. The "unexpected L_DATA_CON" is then the message the state machine was waiting for.

Note: In my opinion the tpuart layer must use interrupts to deal with the UART communication. This would make the receiption much more stable and would avoid loop hogging.

### CODE REVIEW

**Comment by thelsing** [comment]
Merged Thanks.

**Comment by thelsing** [comment]
I'm not sure if you can access the serial interrupts without messing the arduino Serial objects buffers. A better solution would be to use an RTOS and give the knx task a high priority.
